README
Net-STOMP-Client

This module provides an object oriented client interface to interact
with servers supporting STOMP (Streaming Text Orientated Messaging
Protocol). It supports the major features of messaging brokers: SSL,
asynchronous I/O, receipts and transactions.

Note: This module has been initially released on CPAN as Net::STOMP
but has later been renamed to Net::STOMP::Client to avoid a naming
conflict.

INSTALLATION

To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install

To test this module with a STOMP broker, see t/2broker.t.

In addition to the mandatory modules listed in Makefile.PL, the
following optional modules do bring extra functionality:

    Authen::Credential - allows easy authentication support
    IO::Socket::SSL - allows using STOMP over SSL
    Messaging::Message - allows frame <-> message conversions

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Net::STOMP::Client
    perldoc Net::STOMP::Client::Auth
    perldoc Net::STOMP::Client::Connection
    perldoc Net::STOMP::Client::Frame
    perldoc Net::STOMP::Client::HeartBeat
    perldoc Net::STOMP::Client::IO
    perldoc Net::STOMP::Client::Peer
    perldoc Net::STOMP::Client::Receipt
    perldoc Net::STOMP::Client::Version

You can find some examples in the eg directory and a detailed tutorial:

    perldoc Net::STOMP::Client::Tutorial

LICENSE AND COPYRIGHT

Copyright (C) CERN 2010-2013

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


Changes
Revision history for Net-STOMP-Client:

2.2	Thu Nov 28 2013
	* Fixed failover parsing (RT#87025).
	* Improved virtual host documentation in the tutorial (RT#90069).

2.1	Wed Jun  5 2013
	* Now preserving $@ in object destructor.
	* Fixed Net::STOMP::Client::Auth import bug.

2.0	Fri Dec 14 2012
	* Major code refactoring, see README.2.x for more information.
	* Cleanup code to please Perl::Critic.
	* Removed support for blocking I/O, only non-blocking I/O is supported.
	* Now using the No::Worries and Params::Validate modules from CPAN.
	* Added STOMP 1.2 support.
	* Added the queue_message() and send_message() methods for convenience.
	* Added optional authentication support via Authen::Credential.
	* Promoted to 2.0 after successful testing.

1.8	Thu Sep 27 2012
	* Ignore I/O errors after we have successfully sent the DISCONNECT frame.
	* Added support for non-blocking I/O, this is now the default I/O mode.

1.7	Mon Jul  2 2012
	* Improved SSL support by handling situations where select()
	  claims that there is nothing to read although there is.

1.6	Mon Jun 25 2012
	* Net::STOMP::Client::Frame: minor encoding speed improvements.
	* Net::STOMP::Client: the disconnect() method now handles receipts.
	* Preserve $@ when calling Encode::encode() or Encode::decode().

1.5	Tue Feb 28 2012
	* Improved the documentation wrt failover support id (RT#74738).
	* Added a server() method to access the CONNECTED frame's server header.

1.4	Fri Jan 27 2012
	* Improved the documentation wrt subscription id (RT#73060).
	* Improved ActiveMQ failover URI support.

1.3	Wed Nov 23 2011
	* Fixed our use of isa() (RT#71520).
	* Added support for Messaging::Message (this module is optional).
	* Improved error handling.

1.2	Tue Sep 27 2011
	* Net::STOMP::Client::Frame: the frame parsing, decoding and encoding
	  code is now documented and can be used outside of this module.
	* Improved frame debugging to really show what has been sent/received.
	* Net::STOMP::Client::Frame: removed the debug() method.
	* Fixed typo (RT#69417).
	* Improved session id handling in cases where the session header is
	  missing from the CONNECTED frame (it is optional wrt the spec).
	* Net::STOMP::Client: the send() method can now take a body reference.

1.1	Fri Jun 24 2011
	* Net::STOMP::Client::Frame: added a body_reference() method
	  to avoid string copies when manipulating large bodies.
	* Greatly improved the overall performance (~4 times faster),
	  mainly by reducing string copies, improving frame parsing
	  and minimizing syscalls in the I/O layer.
	* Added several methods to Net::STOMP::Client to get better
	  performance, see the "EFFICIENT I/O" section of the tutorial.

1.0	Tue May 10 2011
	* Promoted to 1.0 after successful testing.
	* The API is now stable.
	* The module $VERSION variables now contain the package version
          (RT#67227).

0.9.7	Tue May  3 2011
	* Fixed spurious test errors.
	* Documented how to get strict STOMP 1.1 compliance.

0.9.6	Mon Apr  4 2011
	* Added headers backslash escaping for STOMP 1.1.
	* Added some example scripts in the "eg" directory.
	* Finished STOMP 1.1 support according to final specification.
	* Improved frame checking.
	* Added a socket() method giving access to the socket used.
	* Added support for frame body printing while debugging.

0.9.5	Fri Dec 17 2010
	* Made login and passcode optional in the CONNECT frame.
	* Net::STOMP::Client->wait_for_frames() now uses a high-res timer.
	* Added Net::STOMP::Client::Tutorial.
	* Improved STOMP 1.1 support to match the latest draft specification.

0.9.4	Wed Oct  6 2010
	* Added STOMP 1.1 support (experimental).
	* Changed encoding handling in Net::STOMP::Client::Frame to be
	  compatible with STOMP 1.1
	* Added Net::STOMP::Client::Peer to keep track of peer information.
	* Added Net::STOMP::Client::Protocol to isolate protocol information.
	* Added more tests, some optionally using a broker connection.
	* Improved the documentation.

0.9.3	Tue Jul 20 2010
	* Rewrote the I/O layer to make it more robust.

0.9.2	Mon Jun  7 2010
	* The timeout option now allows the control of different operations
	  (such as waiting for the initial CONNECTED frame or waiting to
	  be able to send an outgoing frame) independently.
	* Added some UTF-8 support, see the Net::STOMP::Client::Frame
	  module documentation for more information.

0.9.1	Tue Jun  1 2010
	* Now performs a better socket cleanup when disconnecting.
	* Net::STOMP::Client->wait_for_frames() now handles more precisely
	  the timeout option: if integer, it should wait for at least this
	  number of seconds and should not block for more than this number
	  of seconds plus one.
	* All STOMP methods now support a timeout option that specifies the
	  maximum time that can be used to send the frame.

0.9	Mon Mar 29 2010
	* Net::STOMP::Client::Frame is now less strict about header names.
	* Net::STOMP::Client->new() now uses its timeout option to set the
	  IO::Socket::INET Timeout option (unless it already exists).

0.8	Mon Mar  1 2010
	* This module has been initially released on CPAN as Net::STOMP
	  but has later been renamed to Net::STOMP::Client to avoid a
	  naming conflict.


lib/Net/STOMP/Client/Protocol.pm
#+##############################################################################
#                                                                              #
# File: Net/STOMP/Client/Protocol.pm                                           #
#                                                                              #
# Description: Protocol support for Net::STOMP::Client                         #
#                                                                              #
#-##############################################################################

#
# module definition
#

package Net::STOMP::Client::Protocol;
use strict;
use warnings;
our $VERSION  = "1.8";
our $REVISION = sprintf("%d.%02d", q$Revision: 1.16 $ =~ /(\d+)\.(\d+)/);

#
# export control
#

use Exporter;
our(@ISA, @EXPORT, @EXPORT_OK, %EXPORT_TAGS);
@ISA = qw(Exporter);
@EXPORT = qw();
@EXPORT_OK = qw();
%EXPORT_TAGS = (
    "CONSTANTS" => [qw(
        ANY_VERSION
        FLAG_BODY_MANDATORY FLAG_BODY_FORBIDDEN FLAG_BODY_ANY FLAG_BODY_MASK
        FLAG_DIRECTION_C2S FLAG_DIRECTION_S2C FLAG_DIRECTION_ANY FLAG_DIRECTION_MASK
        FLAG_FIELD_OPTIONAL FLAG_FIELD_MANDATORY FLAG_FIELD_MASK 
        FLAG_TYPE_UNKNOWN FLAG_TYPE_BOOLEAN FLAG_TYPE_INTEGER FLAG_TYPE_LENGTH FLAG_TYPE_MASK
    )],
    "VARIABLES" => [qw(%CommandFlags %CommandCheck %FieldFlags %FieldCheck)],
);
Exporter::export_tags();

#
# used modules
#

use Net::STOMP::Client::Error;

#
# constants
#

use constant ANY_VERSION => "*";

use constant FLAG_BODY_MANDATORY  => 1;
use constant FLAG_BODY_FORBIDDEN  => 2;
use constant FLAG_BODY_ANY        => 3;
use constant FLAG_BODY_MASK
    => (FLAG_BODY_MANDATORY|FLAG_BODY_FORBIDDEN|FLAG_BODY_ANY);

use constant FLAG_DIRECTION_C2S   => 1 << 2;
use constant FLAG_DIRECTION_S2C   => 2 << 2;
use constant FLAG_DIRECTION_ANY   => 3 << 2;
use constant FLAG_DIRECTION_MASK
    => (FLAG_DIRECTION_C2S|FLAG_DIRECTION_S2C|FLAG_DIRECTION_ANY);

use constant FLAG_FIELD_OPTIONAL  => 1;
use constant FLAG_FIELD_MANDATORY => 2;
use constant FLAG_FIELD_MASK
    => (FLAG_FIELD_OPTIONAL|FLAG_FIELD_MANDATORY);

use constant FLAG_TYPE_UNKNOWN    => 0 << 2;
use constant FLAG_TYPE_BOOLEAN    => 1 << 2;
use constant FLAG_TYPE_INTEGER    => 2 << 2;
use constant FLAG_TYPE_LENGTH     => 3 << 2;
use constant FLAG_TYPE_MASK
    => (FLAG_TYPE_UNKNOWN|FLAG_TYPE_BOOLEAN|FLAG_TYPE_INTEGER|FLAG_TYPE_LENGTH);

#
# shortcuts
#

use constant FLAG_FIELD_OPTIONAL_BOOLEAN => FLAG_FIELD_OPTIONAL | FLAG_TYPE_BOOLEAN;
use constant FLAG_FIELD_OPTIONAL_INTEGER => FLAG_FIELD_OPTIONAL | FLAG_TYPE_INTEGER;
use constant FLAG_FIELD_OPTIONAL_LENGTH  => FLAG_FIELD_OPTIONAL | FLAG_TYPE_LENGTH;

#
# global variables
#

our(
    %CommandFlags,		# version x command => flags
    %CommandCheck,		# version x command => checking code
    %FieldFlags,		# version x command x filed => flags
    %FieldCheck,		# version x command x filed => checking regexp or code
);

#
# references:
#  - STOMP 1.0: http://stomp.github.com/stomp-specification-1.0.html
#  - STOMP 1.1: http://stomp.github.com/stomp-specification-1.1.html
#

#
# known commands for both STOMP 1.0 and STOMP 1.1
#

# client -> server (only SEND can have a body)
foreach my $command (qw(CONNECT SEND SUBSCRIBE UNSUBSCRIBE
                        BEGIN COMMIT ABORT ACK DISCONNECT)) {
    $CommandFlags{ANY_VERSION()}{$command} =
	FLAG_DIRECTION_C2S |
        ($command =~ /^(SEND)$/ ? FLAG_BODY_ANY : FLAG_BODY_FORBIDDEN);
}

# server -> client (only MESSAGE and ERROR can have a body)
foreach my $command (qw(CONNECTED MESSAGE RECEIPT ERROR)) {
    $CommandFlags{ANY_VERSION()}{$command} =
	FLAG_DIRECTION_S2C |
        ($command =~ /^(MESSAGE|ERROR)$/ ? FLAG_BODY_ANY : FLAG_BODY_FORBIDDEN);
}

# STOMP 1.1 extensions
$CommandFlags{"1.1"}{NACK} = FLAG_DIRECTION_C2S | FLAG_BODY_FORBIDDEN;

#
# known fields for both STOMP 1.0 and STOMP 1.1
#

$FieldFlags{ANY_VERSION()}{CONNECT}{"login"} = FLAG_FIELD_OPTIONAL;
$FieldFlags{ANY_VERSION()}{CONNECT}{"passcode"} = FLAG_FIELD_OPTIONAL;

$FieldFlags{ANY_VERSION()}{SEND}{"destination"} = FLAG_FIELD_MANDATORY;
$FieldFlags{ANY_VERSION()}{SEND}{"transaction"} = FLAG_FIELD_OPTIONAL;

$FieldFlags{ANY_VERSION()}{SUBSCRIBE}{"destination"} = FLAG_FIELD_MANDATORY;
# nota bene: ack can only contain some values, this is checked elsewhere
$FieldFlags{ANY_VERSION()}{SUBSCRIBE}{"ack"} = FLAG_FIELD_OPTIONAL;
# id is optional in 1.0 but mandatory in 1.1
$FieldFlags{"1.0"}{SUBSCRIBE}{"id"} = FLAG_FIELD_OPTIONAL;
$FieldFlags{"1.1"}{SUBSCRIBE}{"id"} = FLAG_FIELD_MANDATORY;

# nota bene: in 1.0, either destination or id must be present
$FieldFlags{"1.0"}{UNSUBSCRIBE}{"destination"} = FLAG_FIELD_OPTIONAL;
$FieldFlags{"1.0"}{UNSUBSCRIBE}{"id"} = FLAG_FIELD_OPTIONAL;
# however for 1.1, only id can be used and it is mandatory
$FieldFlags{"1.1"}{UNSUBSCRIBE}{"id"} = FLAG_FIELD_MANDATORY;

$FieldFlags{ANY_VERSION()}{BEGIN}{"transaction"} = FLAG_FIELD_MANDATORY;

$FieldFlags{ANY_VERSION()}{COMMIT}{"transaction"} = FLAG_FIELD_MANDATORY;

$FieldFlags{ANY_VERSION()}{ABORT}{"transaction"} = FLAG_FIELD_MANDATORY;

$FieldFlags{ANY_VERSION()}{ACK}{"message-id"} = FLAG_FIELD_MANDATORY;
$FieldFlags{ANY_VERSION()}{ACK}{"transaction"} = FLAG_FIELD_OPTIONAL;

# DISCONNECT does not have any specific fields

$FieldFlags{ANY_VERSION()}{CONNECTED}{"session"} = FLAG_FIELD_OPTIONAL;

$FieldFlags{ANY_VERSION()}{MESSAGE}{"destination"} = FLAG_FIELD_MANDATORY;
$FieldFlags{ANY_VERSION()}{MESSAGE}{"message-id"} = FLAG_FIELD_MANDATORY;
# subscription is optional in 1.0 but mandatory in 1.1
$FieldFlags{"1.0"}{MESSAGE}{"subscription"} = FLAG_FIELD_OPTIONAL;
$FieldFlags{"1.1"}{MESSAGE}{"subscription"} = FLAG_FIELD_MANDATORY;

$FieldFlags{ANY_VERSION()}{RECEIPT}{"receipt-id"} = FLAG_FIELD_MANDATORY;

$FieldFlags{ANY_VERSION()}{ERROR}{"message"} = FLAG_FIELD_OPTIONAL;

#
# STOMP 1.1 extensions
#

$FieldFlags{"1.1"}{CONNECT}{"accept-version"} = FLAG_FIELD_MANDATORY;
$FieldFlags{"1.1"}{CONNECT}{"host"} = FLAG_FIELD_MANDATORY;
$FieldFlags{"1.1"}{CONNECT}{"heart-beat"} = FLAG_FIELD_OPTIONAL;

$FieldFlags{"1.1"}{CONNECTED}{"version"} = FLAG_FIELD_MANDATORY;
$FieldFlags{"1.1"}{CONNECTED}{"server"} = FLAG_FIELD_OPTIONAL;
$FieldFlags{"1.1"}{CONNECTED}{"heart-beat"} = FLAG_FIELD_OPTIONAL;

$FieldFlags{"1.1"}{ACK}{"subscription"} = FLAG_FIELD_MANDATORY;

$FieldFlags{"1.1"}{NACK}{"subscription"} = FLAG_FIELD_MANDATORY;
$FieldFlags{"1.1"}{NACK}{"message-id"} = FLAG_FIELD_MANDATORY;
$FieldFlags{"1.1"}{NACK}{"transaction"} = FLAG_FIELD_OPTIONAL;

# since these are not explicitly flagged as forbidden in 1.0, we flag them all as optional...
foreach my $command (keys(%{ $FieldFlags{"1.1"} })) {
    foreach my $field (keys(%{ $FieldFlags{"1.1"}{$command} })) {
	$FieldFlags{ANY_VERSION()}{$command}{$field} ||= FLAG_FIELD_OPTIONAL;
    }
}

#
# standard STOMP 1.0 or 1.1 fields
#

foreach my $version (keys(%CommandFlags)) {
    foreach my $command (keys(%{ $CommandFlags{$version} })) {
	# any frame can have a content-type header
	$FieldFlags{$version}{$command}{"content-type"} = FLAG_FIELD_OPTIONAL;
	# any frame can have a content-length header which must be an integer
	$FieldFlags{$version}{$command}{"content-length"} = FLAG_FIELD_OPTIONAL_LENGTH;
	# any client frame (except CONNECT) can have a receipt header
	next if ($CommandFlags{$version}{$command} & FLAG_DIRECTION_MASK)
	    == FLAG_DIRECTION_S2C;
	next if $command =~ /^(CONNECT)$/;
	$FieldFlags{$version}{$command}{"receipt"} = FLAG_FIELD_OPTIONAL;
    }
}

#
# widely used extensions
#

foreach my $command (qw(SUBSCRIBE UNSUBSCRIBE SEND MESSAGE)) {
    $FieldFlags{ANY_VERSION()}{$command}{"persistent"} = FLAG_FIELD_OPTIONAL_BOOLEAN;
}
foreach my $command (qw(SEND MESSAGE)) {
    $FieldFlags{ANY_VERSION()}{$command}{"correlation-id"} = FLAG_FIELD_OPTIONAL;
    $FieldFlags{ANY_VERSION()}{$command}{"priority"} = FLAG_FIELD_OPTIONAL_INTEGER;
    $FieldFlags{ANY_VERSION()}{$command}{"reply-to"} = FLAG_FIELD_OPTIONAL;
}

#
# STOMP JMS Bindings (http://stomp.codehaus.org/StompJMS)
#

$FieldFlags{ANY_VERSION()}{SUBSCRIBE}{"selector"} = FLAG_FIELD_OPTIONAL;
$FieldFlags{ANY_VERSION()}{SUBSCRIBE}{"no-local"} = FLAG_FIELD_OPTIONAL_BOOLEAN;
$FieldFlags{ANY_VERSION()}{SUBSCRIBE}{"durable-subscriber-name"} = FLAG_FIELD_OPTIONAL;

#
# STOMP extensions for JMS message semantics (http://activemq.apache.org/stomp.html)
# plus JMSXUserID (http://activemq.apache.org/jmsxuserid.html)
#

foreach my $field (qw(expires type JMSXGroupID JMSXGroupSeq JMSXUserID)) {
    my $type = $field eq "expires" ? FLAG_TYPE_INTEGER : FLAG_TYPE_UNKNOWN;
    $FieldFlags{ANY_VERSION()}{SEND}{$field} = FLAG_FIELD_OPTIONAL | $type;
    $FieldFlags{ANY_VERSION()}{MESSAGE}{$field} = FLAG_FIELD_OPTIONAL | $type;
}

#
# ActiveMQ extensions to STOMP (http://activemq.apache.org/stomp.html)
#

$FieldFlags{ANY_VERSION()}{CONNECT}{"client-id"} = FLAG_FIELD_OPTIONAL;
foreach my $field (qw(dispatchAsync exclusive maximumPendingMessageLimit noLocal
		      prefetchSize priority retroactive subscriptionName)) {
    my $type = FLAG_TYPE_UNKNOWN;
    $type = FLAG_TYPE_BOOLEAN
        if $field =~ /^(dispatchAsync|exclusive|noLocal|retroactive)$/;
    $type = FLAG_TYPE_INTEGER
        if $field =~ /^(maximumPendingMessageLimit|prefetchSize|priority)$/;
    $FieldFlags{ANY_VERSION()}{SUBSCRIBE}{"activemq.$field"} = FLAG_FIELD_OPTIONAL | $type;
}

#
# ActiveMQ extensions for advisory messages (http://activemq.apache.org/advisory-message.html)
#

foreach my $field (qw(originBrokerId originBrokerName originBrokerURL orignalMessageId
		      consumerCount producerCount consumerId producerId usageName)) {
    $FieldFlags{ANY_VERSION()}{MESSAGE}{$field} = FLAG_FIELD_OPTIONAL;
}

#
# Apollo extensions to STOMP (http://activemq.apache.org/apollo/versions/1.1-SNAPSHOT/website/documentation/user-manual.html)
#

$FieldFlags{ANY_VERSION()}{CONNECTED}{"user-id"} = FLAG_FIELD_OPTIONAL;
$FieldFlags{ANY_VERSION()}{MESSAGE}{"browser"} = FLAG_FIELD_OPTIONAL;
$FieldFlags{ANY_VERSION()}{SEND}{"retain"} = FLAG_FIELD_OPTIONAL;
$FieldFlags{ANY_VERSION()}{SUBSCRIBE}{"browser"} = FLAG_FIELD_OPTIONAL_BOOLEAN;
$FieldFlags{ANY_VERSION()}{SUBSCRIBE}{"browser-end"} = FLAG_FIELD_OPTIONAL_BOOLEAN;
$FieldFlags{ANY_VERSION()}{SUBSCRIBE}{"credit"} = FLAG_FIELD_OPTIONAL;
$FieldFlags{ANY_VERSION()}{SUBSCRIBE}{"exclusive"} = FLAG_FIELD_OPTIONAL_BOOLEAN;
$FieldFlags{ANY_VERSION()}{SUBSCRIBE}{"from-seq"} = FLAG_FIELD_OPTIONAL_INTEGER;
$FieldFlags{ANY_VERSION()}{SUBSCRIBE}{"include-seq"} = FLAG_FIELD_OPTIONAL;

#
# RabbitMQ extensions to STOMP (http://www.rabbitmq.com/stomp.html)
#

$FieldFlags{ANY_VERSION()}{SUBSCRIBE}{"prefetch-count"} = FLAG_FIELD_OPTIONAL_INTEGER;
foreach my $field (qw(amqp-message-id content-encoding)) {
    $FieldFlags{ANY_VERSION()}{SEND}{$field} = FLAG_FIELD_OPTIONAL;
    $FieldFlags{ANY_VERSION()}{MESSAGE}{$field} = FLAG_FIELD_OPTIONAL;
}

#
# other undocumented headers :-(
#

$FieldFlags{ANY_VERSION()}{MESSAGE}{"receipt"} = FLAG_FIELD_OPTIONAL;
$FieldFlags{ANY_VERSION()}{MESSAGE}{"redelivered"} = FLAG_FIELD_OPTIONAL;
$FieldFlags{ANY_VERSION()}{MESSAGE}{"timestamp"} = FLAG_FIELD_OPTIONAL;
$FieldFlags{ANY_VERSION()}{MESSAGE}{"JMSXMessageCounter"} = FLAG_FIELD_OPTIONAL;

#
# additional checking specifications
#

$CommandCheck{"1.0"}{UNSUBSCRIBE} = sub ($$$) {
    my($command, $headers, $body) = @_;

    # either destination or id must be given
    unless ($headers->{destination} or $headers->{id}) {
	Net::STOMP::Client::Error::report("%s: missing header key for %s: destination or id",
					  "Net::STOMP::Client::Protocol::check", $command);
	return();
    }
    return(1);
};

$FieldCheck{"1.0"}{SUBSCRIBE}{"ack"} = qr/^(auto|client)$/;
$FieldCheck{"1.1"}{SUBSCRIBE}{"ack"} = qr/^(auto|client|client-individual)$/;

$FieldCheck{"1.1"}{CONNECT}{"heart-beat"} = qr/^\d+,\d+$/;
$FieldCheck{"1.1"}{CONNECTED}{"heart-beat"} = qr/^\d+,\d+$/;

1;

__END__

=head1 NAME

Net::STOMP::Client::Protocol - Protocol support for Net::STOMP::Client

=head1 DESCRIPTION

This module provides protocol support for Net::STOMP::Client. It
contains global variables defining how STOMP frames should look like.

It is used internally by Net::STOMP::Client::Frame and is not expected
to be used elsewhere.

=head1 AUTHOR

Lionel Cons L<http://cern.ch/lionel.cons>

Copyright CERN 2010-2012
